### PR TITLE
NETOBSERV-334 concurrent map iteration & write fix

### PR DESCRIPTION
**Pull request by jpinsonneau**
This fix `fatal error: concurrent map iteration and map write` when calling `stdout` and `loki` stages in parallel

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -118,3 +118,14 @@ func ParseConfig() error {
...
+// Copy will create a flat copy of GenericMap
+func (m GenericMap) Copy() map[string]interface{} {
+	result := map[string]interface{}{}
+
+	for k, v := range m {
+		result[k] = v
+	}
+
+	return result
+}
```
Comment (jpinsonneau) [inline pkg/config/config.go:131]: I do a simple flat copy here for performances purpose but we can implement a deep copy if needed

**Review by mariomac — APPROVED** [review]

File: pkg/config/config.go (hunk 1 above)
Comment (KalmanMeth) [reply pkg/config/config.go:131]: Should this function be here or in utils?

File: pkg/config/config.go [hunk 2]
```diff
@@ -118,3 +118,14 @@ func ParseConfig() error {
 	logrus.Debugf("params = %v ", Parameters)
 	return nil
 }
+
+// Copy will create a flat copy of GenericMap
+func (m GenericMap) Copy() map[string]interface{} {
```
Comment (eranra) [inline pkg/config/config.go:123]: Should the input and output be both `GenericMap` ??? 

i.e. s/ `map[string]interface{}` / `GenericMap`

File: pkg/config/config.go (hunk 1 above)
Comment (eranra) [reply pkg/config/config.go:131]: @KalmanMeth @jpinsonneau I had the same thought .... maybe the right thing to do is open a new GenericMap.go file

File: pkg/pipeline/write/write_loki.go [hunk 3]
```diff
@@ -231,7 +231,8 @@ func (l *Loki) processRecords() {
 			log.Debugf("exiting writeLoki because of signal")
 			return
 		case record := <-l.in:
-			err := l.ProcessRecord(record)
+			// copy record before process to avoid alteration on parallel stages
+			err := l.ProcessRecord(record.Copy())
```
Comment (eranra) [inline pkg/pipeline/write/write_loki.go:235]: @jpinsonneau I suggest moving the copy process to be inside the `ProcessRecord` function --- the rule should be that we are not allowed to change the inputs --- and we need to maybe check also in other places that we are not doing that.

**Review by eranra — APPROVED** [review]
Some suggestions for improvement but from my PoV this can merge as-is. so LGTM

File: pkg/config/config.go (hunk 1 above)
Comment (jpinsonneau) [reply pkg/config/config.go:131]: I can move this :+1: 

Does a new file `generic_map.go` in config folder is ok for both of you ? @KalmanMeth @eranra

File: pkg/config/config.go (hunk 2 above)
Comment (jpinsonneau) [reply pkg/config/config.go:123]: sounds good :+1:

File: pkg/config/config.go (hunk 1 above)
Comment (jpinsonneau) [reply pkg/config/config.go]: https://github.com/netobserv/flowlogs-pipeline/pull/226/commits/2648b82480b2f4a2515e3bee5055568afa7a47be#diff-2eff304122c33ebdc65982528ca9efc9fd8061e6c75fda0beffd1b3ce0cdc046

File: pkg/pipeline/write/write_loki.go (hunk 3 above)
Comment (eranra) [reply pkg/pipeline/write/write_loki.go]: @KalmanMeth ... so should we add copy also there ??? do you want to open a PR for that?

**Review by KalmanMeth — APPROVED** [review]

**Review by ronensc — APPROVED** [review]

**Comment by jpinsonneau** [comment]
Thank you all for the feedback :+1:

**Pull request merged**